xdsl/irdl/declarative_assembly_format.py

```diff
@@ -207,14 +203,22 @@ def resolve_operand_types(self, state: ParsingState, op_def: OpDef) -> None:
             zip(state.operand_types, op_def.operands, strict=True)
         ):
             if operand_type is None:
-                operand_type = operand_def.constr.infer(state.constraint_variables)
                 operand = state.operands[i]
-                if isinstance(operand, UnresolvedOperand):
-                    state.operand_types[i] = operand_type
-                elif isinstance(operand, list):
-                    state.operand_types[i] = cast(
-                        list[Attribute | None], [operand_type]
-                    ) * len(operand)
+                range_length = len(operand) if isinstance(operand, list) else 1
+                operand_type = operand_def.constr.infer(
+                    range_length, state.constraint_variables
+                )
+                if isinstance(operand_def, OptionalDef):
+                    operand_type = (
+                        list[Attribute | None]()
+                        if len(operand_type) == 0
+                        else operand_type[0]
+                    )
+                elif isinstance(operand_def, VariadicDef):
+                    operand_type = cast(list[Attribute | None], operand_type)
```

Comment:
also surprised you need the cast here

Reply:
and below

Reply:
`Argument of type "list[Attribute | None] | Attribute | list[Attribute]" cannot be assigned to parameter "value" of type "Attribute | list[Attribute | None] | None"` ?